hpy/devel/include/hpy/cpy_types.h

```diff
@@ -1,15 +1,42 @@
 #ifndef HPY_UNIVERSAL_CPY_TYPES_H
 #define HPY_UNIVERSAL_CPY_TYPES_H
 
-/* generally speaking, we can't #include Python.h, but there are a bunch of
- * types defined there that we need to use.  Here, we redefine all the types
- * we need, with a cpy_ prefix.
- */
+/* ~~~~~~~~~~~~~~~~ CPython legacy types ~~~~~~~~~~~~~~~~ */
+
+/* These are the types which are needed to implement legacy features such as
+   .legacy_slots, .legacy_methods, HPy_FromPyObject, HPy_AsPyObject, etc.
+
+   In cpython and hybrid ABI mode we can #include Python.h and use the "real"
+   types.
+
+   In universal ABI mode, legacy features cannot be used, but we still need
+   the corresponding C types to use in the HPy declarations. Note that we use
+   only forward declarations, meaning that it will actually be impossible to
+   instantiate any of these struct.
+*/
+
+#ifdef HPY_ABI_UNIVERSAL
+
+typedef struct FORBIDDEN_cpy_PyObject cpy_PyObject;
+typedef struct FORBIDDEN_PyMethodDef cpy_PyMethodDef;
+typedef struct FORBIDDEN_bufferinfo cpy_Py_buffer;
+
+// declare the following API functions as _HPY_LEGACY, which triggers an
+// #error if they are used
+HPyAPI_FUNC _HPY_LEGACY cpy_PyObject *HPy_AsPyObject(HPyContext *ctx, HPy h);
+HPyAPI_FUNC _HPY_LEGACY HPy HPy_FromPyObject(HPyContext *ctx, cpy_PyObject *obj);
+
+#else
+
+// Python.h has already been included by the main hpy.h
+typedef PyObject cpy_PyObject;
+typedef PyMethodDef cpy_PyMethodDef;
+typedef Py_buffer cpy_Py_buffer;
```

Comment:
Ah, that's great and will fix our `CPython >=3.11` compatibility problem.
Maybe you could mention issue #288 in this PR.

Reply:
right! I edited the description to mention it